### PR TITLE
Emit more events from Scene3D

### LINKED ISSUE #209: Send more events from Scene3D
## Desired behavior

Since https://github.com/ignitionrobotics/ign-gui/issues/68, any plugin can [safely interact with the 3D scene during the render event](https://ignitionrobotics.org/api/gazebo/5.0/rendering_plugins.html). We should expose more events from the rendering scene, so plugins can react to clicks, hover, etc.

Ignition Gazebo's `GzScene3D` plugin already [broadcasts right/left clicks and hover events](https://github.com/ignitionrobotics/ign-gazebo/blob/26832764749291b6ea22b525262136ecbb63fbb9/src/gui/plugins/scene3d/Scene3D.cc#L989-L1036) using Ignition GUI events. We just need to add that functionality to `ign-gui`'s `Scene3D`.

## Alternatives considered

One alternative that users have now is forking `Scene3D` and embedding functionality into it, but that doesn't scale well and doesn't allow multiple plugins from various developers to play well together.

## Implementation suggestion

We can do exactly the same as [the ign-gazebo implementation](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo5/src/gui/plugins/scene3d/Scene3D.cc#L989-L1036).

## Additional context

This should help with https://github.com/ignitionrobotics/ign-rviz/issues/70.

### PR DESCRIPTION
**Pull request by chapulina**
# 🎉 New feature

Closes #209

Related to https://github.com/ignitionrobotics/ign-rviz/issues/70

## Summary

This PR makes the 3D scene emit more events, such as clicks and hover, that can be used by other plugins. Most of the code here has been ported from [ignition::gazebo::Scene3D](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo5/src/gui/plugins/scene3d/Scene3D.hh).

I got stuck writing tests for it though, for some reason the render engine plugin is not properly loaded. I spent some time digging deep into `ign-plugin` with no luck. I'll leave this as draft until I can address this.

## Test it

I'd like to add an example plugin that makes use of these events, so others can use it as a pattern to copy for their needs.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests - **that's the tough part**
- [ ] Added example and/or tutorial - **TODO**
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

**Comment by chapulina** [comment]
I removed the tests that weren't working and ticketed #216. Opening this for review.

It's worth reading the conversation on #209. The current approach to emitting events is necessary for some use cases, but adds unnecessary complexity for others. I think we could revisit the current architecture in a future release in favor of an `objectName`-based approach for specific use cases.

**Comment by scpeters** [comment]
I've fixed the issue with loading the render engines documented in #216 and have merged those changes into this branch. I moved `Scene3D_TEST.cc` to `test/integration`, and it seems to be working now.

**Comment by chapulina** [comment]
I added a test for the `render` event in 23c0578. I won't have time to add tests for the other events for a while. Options:

1. Get this in with the current tests and add more tests later (this is already more testing than most plugins)
2. Someone else picks up this PR and adds tests for the click and hover events
3. Wait until I come back

**Comment by scpeters** [comment]
> I added a test for the `render` event in [23c0578](https://github.com/ignitionrobotics/ign-gui/commit/23c057865ed3f50268bd4a94f6ac57c03fb25973). I won't have time to add tests for the other events for a while. Options:
> 
> 1. Get this in with the current tests and add more tests later (this is already more testing than most plugins)
> 2. Someone else picks up this PR and adds tests for the click and hover events
> 3. Wait until I come back

I can try adding some more tests. Do you know how to create fake mouse clicks from a test?

**Comment by scpeters** [comment]
I just noticed the windows build has been broken since 23c0578

~~~
error C2491: 'ignition::gui::TestHelper::staticMetaObject': definition of dllimport static data member not allowed
~~~

**Comment by scpeters** [comment]
I've added some expectations about the value of the `Vector3d` passed with the click events in 48171dd, and if CI is reasonable, then I will be ready to approve

File: test/integration/scene3d.cc [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+      receivedLeftEvent = true;
+      auto leftClickToScene = static_cast<events::LeftClickToScene*>(_event);
+      leftClickPoint = leftClickToScene->Point();
+    }
+    else if (_event->type() == events::HoverToScene::kType)
+    {
+      receivedHoverEvent = true;
+      auto hoverToScene = static_cast<events::HoverToScene*>(_event);
+      std::cerr << "hoverToScene point["
+                << hoverToScene->Point() << "]" << std::endl;
```
Comment (ahcorde) [inline test/integration/scene3d.cc:193]: Remove this trace ?

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: test/integration/scene3d.cc (hunk 1 above)
Comment (scpeters) [reply test/integration/scene3d.cc]: 039ef43

**Review by scpeters — APPROVED** [review]

**Review by ahcorde — APPROVED** [review]

**Pull request merged**